File: solutions/lua/square-root/4/square-root.lua
local SquareRoot = {}

local function square_root(radicand)
	if radicand < 2 then
		return radicand
	end

	local smaller = square_root(radicand >> 2) << 1
	local larger = smaller + 1

	if larger * larger > radicand then
		return smaller
	else
		return larger
	end
end

SquareRoot.square_root = square_root

return SquareRoot

File: solutions/elixir/square-root/4/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand) do
    Stream.iterate(1, &div(&1 + div(radicand, &1), 2))
    |> Stream.drop_while(&(&1 * &1 != radicand))
    |> Stream.take(1)
    |> Enum.to_list()
    |> List.first()
  end
end
